### PR TITLE
avoid overwriting params with falsey value

### LINKED ISSUE #1417: Inconsistency in setting params from optional route captures between 1.4 and 2.0
I'm wary to call this a bug just yet, but I wanted to bring this up since I'm noticing a different behavior. This kinda intersects with third-party Rack middleware in [`rack-contrib`](https://github.com/rack/rack-contrib) but I'm hoping this is relevant here.

In Sinatra 1.4, it looks like when the requested route has optional parameters in it, the parsed value is added to `params` [when the value is truthy (not `false`/`nil`)](https://github.com/sinatra/sinatra/blob/v1.4.8/lib/sinatra/base.rb#L1010)(excluding the array case). So if there are items already in `params`, like from a Rack middleware such as [PostBodyContentTypeParser](https://github.com/rack/rack-contrib/blob/master/lib/rack/contrib/post_body_content_type_parser.rb), a lack of a match on the optional parameter in the route would not overwrite a value in `params` with `nil.

An example (Sinatra 1.4):

```ruby
app.post '/example/?:id?' do
  # code here
end
```
If I were to hit this via `/example`, `params[:id]` would be `nil` (on account of it not being set during route-optionals-matching). If I were to hit this via `/example/2`, `params[:id] == 2`. If I were to hit this via `/example` and include in my POST body the JSON: `{"id": 2}`, `PostBodyContentTypeParser` would set `params[:id]` to 2 as well, and Sinatra 1.4's optionals-in-route handling would not overwrite with `nil`.

Meanwhile, in Sinatra 2.0, it looks like the resulting matches from the `Mustermann` pattern are [directly merged to the `params` hash if `params.any?` returns something truthy.](https://github.com/sinatra/sinatra/blob/v2.0.1/lib/sinatra/base.rb#L1026) This, _I think_ has the effect of, basically (contrived `irb` example incoming):

```ruby
irb(main):001:0> require 'mustermann/sinatra'
=> true
irb(main):002:0> require 'sinatra/indifferent_hash'
=> true
irb(main):003:0> params = Sinatra::IndifferentHash.new
=> {}
irb(main):004:0> params.merge(id: 1 ) # taking in rack's params
=> {"id"=>1}
irb(main):005:0> pattern = Mustermann::Sinatra.new('/example(/:id)?')
=> #<Mustermann::Sinatra:"/example(/:id)?">
irb(main):006:0> params.merge(pattern.params('/example')) # This is going to override rack preset params with nil
=> {"id"=>nil}
```
So the `params` set by the middleware are lost if they are the same key as the optional in the route.

Was I relying on behavior that was never meant to be stable? Or is this an underlying issue?

And of course, thanks Sinatra team for a wonderful gem. :heart:

### PR DESCRIPTION
**Pull request by namusyaka**
This behavior arises from the commit that mustermann was introduced.
see: https://github.com/sinatra/sinatra/commit/7a9ffccdafb71fef8d33c110d03ca0654dbb9316

That breaks backward compatibility with Sinatra v1.4.X and is not intended behavior.
This changes it so that it overwrites params only if there is a possible value as a block argument.

Fixes #1417

I want a cautious review on this PR. After sufficient review, I'm planning to merge the commit.
/cc @rkh @zzak @nickpelone

### CODE REVIEW

**Comment by namusyaka** [comment]
To fix broken tests, applied #1419. Force pushed, but changes aren't nothing.

**Comment by nickpelone** [comment]
:+1: thanks again!

**Comment by namusyaka** [comment]
@nickpelone Thanks for confirming my patch. I'll let @zzak and @rkh review this. Please wait for merging.

**Comment by rkh** [comment]
The change looks good for what it wants to do.

But I don't think we want this change, it might be opening up for security issue or at least bugs where someone is assuming a param can only be set via pattern (ie, imagine someone limiting what `:id` can capture and then relying on that as enough security):

``` ruby
set :pattern, capture: { ext: %w[home imprint] }

get '/:page?' do
  page = params[:page] || 'home'
  path = File.expand_path(page, page_directory)
  send_file path
end
```

**Comment by nickpelone** [comment]
I've never used the feature you describe above - I sanitize my params (wherever they originated in the stack: optional captures, request bodies, etc) with an external gem.

Without getting too in the weeds here, my approach worked in Sinatra 1.4, but is broken in 2.x, where one of the stated desires was [for people to file a ticket if their app stopped working.](http://zzak.io/log/2016-04-18-the-road-to-sinatra-2.0.html) (and for what it's worth, I imagine the situation described above would happen in Sinatra 1.4.x too, re: assuming captures could only come from the route matcher).

I could even live with an option I had to toggle on, but for this behavior to go totally away, forever, would be very depressing to me. (In addition to necessitating a revision of my API I design / deciding to backport to the entire Sinatra 1.x / Rack 1.x ecosystem and live in the past...eww!)

I have strong interest in seeing this functionality restored - so if another approach is needed I could dive into the code and propose one, failing all else.

Thanks.

**Comment by namusyaka** [comment]
Since received @rkh's comment, I have reconsidered this issue.
As a result, I thought this behavior was certainly undesirable.
This is because even if there are duplicate parameters in path and request body (or query), they should be handled separately.
For example, unlike what is included in the request body, parameters included in path do validation by URI at the previous stage, the request body is not validated.

What I'd like to say is that it is not important that validation is carried out, but the two elements are different in nature.
In other words, the original behavior is wrong, I think.

However, we should look to the fact that this errant behavior was working with traditional sinatra.
For example, we can explicitly write in the document that this behavior has been fixed.

/cc @nickpelone @rkh 

In any case, I decided to close this PR as an unwanted change.
Last, @nickpelone Thank you so much for bringing this up.

**Comment by nickpelone** [comment]
☹️

**Comment by nickpelone** [comment]
Upon further reflection, I figure I can make this "lax" style I was using work with an extra helper method in those route blocks so I guess I'll stop being so grumpy. Thanks for taking a look.
